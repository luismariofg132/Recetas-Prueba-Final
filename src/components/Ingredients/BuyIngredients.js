import React, { useEffect, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { listIngredientAsyn } from '../../redux/actions/IngredientsActions'
import Button from 'react-bootstrap/Button'
import { useNavigate } from 'react-router-dom'
import Swal from 'sweetalert2'
const BuyIngredients = () => {

    const [total, settotal] = useState(0)
    const Navigate = useNavigate()
    const dispatch = useDispatch()

    useEffect(function list() {
        dispatch(listIngredientAsyn())
    }, [dispatch])

    const { ingredient } = useSelector(store => store.ingredient)

    const checked = ({ target }) => {
        const checked = target.checked
        const price = target.value
        if (checked) {
            if (total >= 0) {
                settotal(total + Number(price))
            }
        }
        else {
            settotal(total - Number(price))
        }
    }

    const SeleccionarTodo = () => {
        const checbox = document.getElementsByName('checbox')
        for (let i = 0, n = checbox.length; i < n; i++) {
            checbox[i].checked = checked;
        }
        const totalPay = ingredient.reduce((sum, value) => (typeof Number(value.price) == "number" ? sum + Number(value.price) : sum), 0)
        settotal(totalPay)
    }

    const deseleccionarTodo = () => {
        const checbox = document.getElementsByName('checbox')
        for (let i = 0, n = checbox.length; i < n; i++) {
            checbox[i].checked = !checked;
        }
        settotal(0)
    }

    const pagoProducts = () => {
        Swal.fire(
            'Gracias por comprar',
            'Te esperamos para otra compra',
            'success'
        )
        deseleccionarTodo()
    }

    return (
        <div>
            <div className='ingredientCont'>
                <div>
                    <span>Ingredientes</span>
                    <h4>Risotto de setas (vegano)</h4>
                </div>
                <Button variant="link"
                    onClick={() => SeleccionarTodo()}
                >Seleccionar Todo</Button>
                <Button variant="link"
                    onClick={() => deseleccionarTodo()}
                >Deseleccionar Todo</Button>
                <hr />
                {
                    ingredient?.map((ing, index) => (
                        <div key={index}>
                            <div className='ingredient'>
                                <div>
                                    <input type="checkbox" name="checbox"
                                        onClick={checked}
                                        value={ing.price}
                                    />
                                    <h5>{ing.produc}</h5>
                                    <span>{ing.brand}</span>
                                    <span>{ing.quantity}</span>
                                </div>
                                <div>
                                    <h4 className='price'>$ {ing.price}</h4>
                                </div>
                            </div>
                            <hr />
                        </div>
                    ))
                }
                <div className="d-grid gap-2">
                    <Button variant="success"
                        onClick={() => pagoProducts()}
                    >Pagar $ {total}</Button>
                </div>
                <div className='d-grid gap-2 mt-3'>
                    <Button variant="outline-dark"
                        onClick={() => { Navigate('/ingredients') }}
                    >Editar Ingredientes</Button>
                </div>
            </div>
        </div>
    )
}

export default BuyIngredients