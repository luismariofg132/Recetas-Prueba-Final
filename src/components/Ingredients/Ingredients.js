import React, { useEffect, useState } from 'react'
import Table from 'react-bootstrap/Table'
import Button from 'react-bootstrap/Button'
import Modal from 'react-bootstrap/Modal'
import Form from 'react-bootstrap/Form'
import { useForm } from '../../hooks/useForm'
import { useDispatch, useSelector } from 'react-redux'
import { AddIngredientAsyn, deleteIngredientAsyn, editIngredientAsyn, listIngredientAsyn } from '../../redux/actions/IngredientsActions'
import { useNavigate } from 'react-router-dom'

const Ingredients = () => {
    const [show, setShow] = useState(false)
    const [show2, setShow2] = useState(false)
    const Navigate = useNavigate()
    const dispatch = useDispatch()

    useEffect(function list() {
        dispatch(listIngredientAsyn())
    }, [dispatch])

    const { ingredient } = useSelector(store => store.ingredient)

    const [values, handleInputChange, reset, setValues] = useForm({
        product: "",
        brand: "",
        items: "",
        quantity: "",
        price: "",
        id: ""
    })

    const { product, brand, items, quantity, price, id } = values

    const nuevoIngredient = (e) => {
        e.preventDefault()
        const ingredient = {
            product, brand, items, quantity, price
        }
        dispatch(AddIngredientAsyn(ingredient))
        setShow(false)
        reset()
    }

    const editIngredient = (e) => {
        e.preventDefault()
        const ingredient = {
            product, brand, items, quantity, price
        }
        dispatch(editIngredientAsyn(id, ingredient))
        setShow2(false)
        reset()
    }

    return (
        <div>
            <div className='tableIngredients'>
                <Table striped bordered hover>
                    <thead>
                        <tr>
                            <th>Product</th>
                            <th>Brand</th>
                            <th>Items</th>
                            <th>Quantity</th>
                            <th>Price</th>
                            <th>Action</th>
                        </tr>
                    </thead>
                    <tbody>
                        {
                            ingredient?.map((ing, index) => (
                                <tr key={index}>
                                    <td>{ing.product}</td>
                                    <td>{ing.brand}</td>
                                    <td>{ing.items}</td>
                                    <td>{ing.quantity}</td>
                                    <td>{ing.price}</td>
                                    <td>
                                        <Button variant="dark" onClick={() => {
                                            setShow2(true)
                                            setValues({
                                                product: ing.product,
                                                brand: ing.brand,
                                                items: ing.items,
                                                quantity: ing.quantity,
                                                price: ing.price,
                                                id: ing.id
                                            })
                                        }}
                                        >Edit</Button>
                                        <Button variant="dark"
                                            onClick={() => {
                                                dispatch(deleteIngredientAsyn(ing.id))
                                            }}
                                        >Delete</Button>
                                    </td>
                                </tr>
                            ))
                        }
                    </tbody>
                </Table>
                <div className="d-grid gap-2">
                    <Button variant="dark" onClick={() => setShow(true)}>Agregar Nuevo Ingrediente</Button>
                </div>
                <div className="d-grid gap-2 mt-2">
                    <Button variant="dark" onClick={() => Navigate('/')}>Volver a Comprar</Button>
                </div>
            </div>
            <div>
                <Modal
                    show={show}
                    onHide={() => setShow(false)}
                    dialogClassName="modal-90w"
                    aria-labelledby="example-custom-modal-styling-title"
                >
                    <Modal.Header closeButton>
                        <Modal.Title id="example-custom-modal-styling-title">
                            Agregar Nuevo Ingrediente
                        </Modal.Title>
                    </Modal.Header>
                    <Modal.Body>
                        <form onSubmit={nuevoIngredient}>
                            <div>
                                <Form.Label>Product</Form.Label>
                                <Form.Control
                                    type="text"
                                    name='product'
                                    onChange={handleInputChange}
                                    value={product}
                                    required
                                />
                            </div>
                            <div>
                                <Form.Label>Brand</Form.Label>
                                <Form.Control
                                    type="text"
                                    name='brand'
                                    onChange={handleInputChange}
                                    value={brand}
                                    required
                                />
                            </div>
                            <div>
                                <Form.Label>Items</Form.Label>
                                <Form.Control
                                    type="number"
                                    name='items'
                                    onChange={handleInputChange}
                                    value={items}
                                    required
                                />
                            </div>
                            <div>
                                <Form.Label>Quantity</Form.Label>
                                <Form.Control
                                    type="text"
                                    name='quantity'
                                    onChange={handleInputChange}
                                    value={quantity}
                                    required
                                />
                            </div>
                            <div>
                                <Form.Label>Price</Form.Label>
                                <Form.Control
                                    type="number"
                                    name='price'
                                    onChange={handleInputChange}
                                    value={price}
                                    required
                                />
                            </div>
                            <div className="d-grid gap-2 mt-3">
                                <Button variant="dark" type='submit'>Agregar</Button>
                            </div>
                        </form>
                    </Modal.Body>
                </Modal>
            </div>
            <div>
                <Modal
                    show={show2}
                    onHide={() => {
                        setShow2(false)
                        reset()
                    }}
                    dialogClassName="modal-90w"
                    aria-labelledby="example-custom-modal-styling-title"
                >
                    <Modal.Header closeButton>
                        <Modal.Title id="example-custom-modal-styling-title">
                            Editar Ingrediente
                        </Modal.Title>
                    </Modal.Header>
                    <Modal.Body>
                        <form onSubmit={editIngredient}>
                            <div>
                                <Form.Label>Product</Form.Label>
                                <Form.Control
                                    type="text"
                                    name='product'
                                    onChange={handleInputChange}
                                    value={product}
                                    required
                                />
                            </div>
                            <div>
                                <Form.Label>Brand</Form.Label>
                                <Form.Control
                                    type="text"
                                    name='brand'
                                    onChange={handleInputChange}
                                    value={brand}
                                    required
                                />
                            </div>
                            <div>
                                <Form.Label>Items</Form.Label>
                                <Form.Control
                                    type="number"
                                    name='items'
                                    onChange={handleInputChange}
                                    value={items}
                                    required
                                />
                            </div>
                            <div>
                                <Form.Label>Quantity</Form.Label>
                                <Form.Control
                                    type="text"
                                    name='quantity'
                                    onChange={handleInputChange}
                                    value={quantity}
                                    required
                                />
                            </div>
                            <div>
                                <Form.Label>Price</Form.Label>
                                <Form.Control
                                    type="number"
                                    name='price'
                                    onChange={handleInputChange}
                                    value={price}
                                    required
                                />
                            </div>
                            <div className="d-grid gap-2 mt-3">
                                <Button variant="dark" type='submit'>Editar</Button>
                            </div>
                        </form>
                    </Modal.Body>
                </Modal>
            </div>
        </div>
    )
}

export default Ingredients