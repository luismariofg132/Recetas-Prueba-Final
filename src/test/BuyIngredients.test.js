import { shallow } from 'enzyme'
import BuyIngredients from '../components/Ingredients/BuyIngredients'
import Ingredients from '../components/Ingredients/Ingredients'
import { BrowserRouter as Router, Route, Routes } from 'react-router-dom'
import PrivateRouter from '../Routers/PrivateRouter'

describe('Prueba de compra de ingredientes', () => {
    test('Verifica el permiso de autenticacion para BuyIngredient', () => {
        const isAuthenticated = false
        const wrapper = shallow(
            <Router>
                <Routes>
                    <Route path='/*' element={<PrivateRouter isAuthenticated={isAuthenticated}>
                        <BuyIngredients />
                    </PrivateRouter>} />
                </Routes>
            </Router>)
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find(Route).prop('element').props.isAuthenticated).toBeFalsy()
    })
    test('Verifica el permiso de autenticacion para Ingredient', () => {
        const isAuthenticated = false
        const wrapper = shallow(
            <Router>
                <Routes>
                    <Route path='/*' element={<PrivateRouter isAuthenticated={isAuthenticated}>
                        <Ingredients />
                    </PrivateRouter>} />
                </Routes>
            </Router>)
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find(Route).prop('element').props.isAuthenticated).toBeFalsy()
    })
})