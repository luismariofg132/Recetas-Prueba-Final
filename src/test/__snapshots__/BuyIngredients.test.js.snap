// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba de compra de ingredientes Verifica el permiso de autenticacion para BuyIngredient 1`] = `
<Router
  location={
    Object {
      "hash": "",
      "key": "default",
      "pathname": "/",
      "search": "",
      "state": null,
    }
  }
  navigationType="POP"
  navigator={
    Object {
      "action": "POP",
      "back": [Function],
      "block": [Function],
      "createHref": [Function],
      "forward": [Function],
      "go": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Routes>
    <Route
      element={
        <PrivateRouter
          isAuthenticated={false}
        >
          <BuyIngredients />
        </PrivateRouter>
      }
      path="/*"
    />
  </Routes>
</Router>
`;

exports[`Prueba de compra de ingredientes Verifica el permiso de autenticacion para Ingredient 1`] = `
<Router
  location={
    Object {
      "hash": "",
      "key": "default",
      "pathname": "/",
      "search": "",
      "state": null,
    }
  }
  navigationType="POP"
  navigator={
    Object {
      "action": "POP",
      "back": [Function],
      "block": [Function],
      "createHref": [Function],
      "forward": [Function],
      "go": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Routes>
    <Route
      element={
        <PrivateRouter
          isAuthenticated={false}
        >
          <Ingredients />
        </PrivateRouter>
      }
      path="/*"
    />
  </Routes>
</Router>
`;
